ext/opcache/jit/zend_jit_ir.c

```diff
@@ -16685,19 +16763,125 @@ static bool zend_jit_may_be_in_reg(const zend_op_array *op_array, zend_ssa *ssa,
+	if (jit->delayed_call_level) {
+		ir_ref call = jit_CALL(jit_FP(jit), call);
+		ir_STORE(jit_CALL(jit_IP(jit), prev_execute_data), jit->delayed_call_level == 1 ? IR_NULL : ir_LOAD_A(call));
+		ir_STORE(call, jit_IP(jit));
+	}
+	zend_jit_push_call_frame(jit, opline, NULL, fbc, 0, 0, checked_stack, ir_CONST_ADDR(fbc), IR_NULL);
+	// yes, we'll temporarily reuse the IP, but we must not handle this via the reuse_ip mechanism - we just reset IP to original in end
+	// the primary issue here is that this code is just executed conditionally while the reuse mechanisms are meant to be global
+	jit->track_last_valid_opline = track_last_valid_opline;
+	jit->use_last_valid_opline = use_last_valid_opline;
+	jit->last_valid_opline = last_valid_opline;
+	jit->reuse_ip = reuse_ip;
+
+	// push all args
+	uint32_t call_num_args = ZEND_FLF_NUM_ARGS(opline->opcode);
+	switch (call_num_args) {
+		case 3: jit_ZVAL_COPY(jit, ZEND_ADDR_MEM_ZVAL(ZREG_RX, EX_NUM_TO_VAR(2)), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op1_data_ref), op1_data_info, 1); ZEND_FALLTHROUGH;
+		case 2: jit_ZVAL_COPY(jit, ZEND_ADDR_MEM_ZVAL(ZREG_RX, EX_NUM_TO_VAR(1)), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op2_ref), op2_info, 1); ZEND_FALLTHROUGH;
+		case 1: jit_ZVAL_COPY(jit, ZEND_ADDR_MEM_ZVAL(ZREG_RX, EX_NUM_TO_VAR(0)), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op1_ref), op1_info, 1);
+	}
+
+	// Make call frame externally visible
+	ir_ref rx = jit_IP(jit);
+	ir_STORE(jit_CALL(rx, prev_execute_data), jit_FP(jit));
+	ir_STORE(jit_EG(current_execute_data), rx);
+
+	jit_observer_fcall_begin(jit, rx, observer_handler);
+
+	// JIT: fbc->internal_function.handler(new_frame, return_value)
+	ir_CALL_2(IR_VOID, ir_CONST_FC_FUNC(fbc->internal_function.handler), rx, res_ref);
+
+	jit_observer_fcall_end(jit, rx, res_ref);
+
+	ir_STORE(jit_EG(current_execute_data), ir_LOAD_A(jit_CALL(rx, prev_execute_data)));
+
+	for (int i = 0; i < call_num_args; ++i) {
+		if (zend_jit_needs_arg_dtor(fbc, i, NULL)) {
+			uint32_t offset = EX_NUM_TO_VAR(i);
+			zend_jit_addr var_addr = ZEND_ADDR_MEM_ZVAL(ZREG_RX, offset);
+
+			jit_ZVAL_PTR_DTOR(jit, var_addr, MAY_BE_ANY|MAY_BE_RC1|MAY_BE_RCN, 0, opline);
+		}
+	}
+
+	ir_ref allocated_path = IR_UNUSED;
+
+	// free the call frame
+	if (JIT_G(trigger) != ZEND_JIT_ON_HOT_TRACE ||
+	    !JIT_G(current_frame) ||
+	    !JIT_G(current_frame)->call ||
+	    !TRACE_FRAME_IS_NESTED(JIT_G(current_frame)->call)) {
+		// JIT: zend_vm_stack_free_call_frame(call);
+		// JIT: if (UNEXPECTED(ZEND_CALL_INFO(call) & ZEND_CALL_ALLOCATED))
+		ir_ref if_allocated = ir_IF(ir_AND_U8(
+			ir_LOAD_U8(ir_ADD_OFFSET(rx, offsetof(zend_execute_data, This.u1.type_info) + 2)),
+			ir_CONST_U8(ZEND_CALL_ALLOCATED >> 16)));
+		ir_IF_TRUE_cold(if_allocated);
+
+		ir_CALL_1(IR_VOID, ir_CONST_FC_FUNC(zend_jit_free_call_frame), rx);
+
+		allocated_path = ir_END();
+		ir_IF_FALSE(if_allocated);
+	}
+	ir_STORE(jit_EG(vm_stack_top), rx);
+
+	if (allocated_path) {
+		ir_MERGE_WITH(allocated_path);
+	}
```

Comment:
Can you use ``zend_jit_do_fcall()``, to minimize the JIT code duplication.

Reply:
This duplicated version doesn't do exactly the same as VM. At least, it misses support for ``zend_execute_internal()``.
It's going to be very difficult to maintain this duplicated code.

Reply:
zend_execute_internal prevents frameless alltogether.

Reply:
In VM you falback FRAMELESS_CALL to DO_ICALL so zend_execute_internal() is going to be called.

Reply:
do this in a separate patch.

Reply:
So, first merge this one, then improve upon this in a later PR? okay.